py-polars/polars/io/cloud/credential_provider.py

```diff
@@ -139,6 +141,149 @@ def _check_module_availability(cls) -> None:
+                    f"{self.scopes = } "
+                ),
+                file=sys.stderr,
+            )
+
+    def __call__(self) -> CredentialProviderFunctionReturn:
+        """Fetch the credentials."""
+        if self.account_name is not None:
+            try:
+                out = (
+                    {
+                        "account_key": self._get_azure_storage_account_key_az_cli(
+                            self.account_name
+                        )
+                    },
+                    None,
+                )
+
+                if self._verbose:
+                    print(
+                        "[CredentialProviderAzure]: retrieved account keys from Azure CLI",
+                        file=sys.stderr,
+                    )
+            except Exception as e:
+                if self._verbose:
+                    print(
+                        f"[CredentialProviderAzure]: failed to retrieve account keys from Azure CLI: {e}",
+                        file=sys.stderr,
+                    )
+            else:
+                return out  # type: ignore[return-value]
+
+        token = self.credential.get_token(*self.scopes)
+
+        return {
+            "bearer_token": token.token,
+        }, token.expires_on
+
+    @classmethod
+    def _check_module_availability(cls) -> None:
+        if importlib.util.find_spec("azure.identity") is None:
+            msg = "azure-identity must be installed to use `CredentialProviderAzure`"
+            raise ImportError(msg)
+
+    @staticmethod
+    def _extract_adls_uri_storage_account(uri: str) -> str | None:
+        # "abfss://{CONTAINER}@{STORAGE_ACCOUNT}.dfs.core.windows.net/"
+        #                      ^^^^^^^^^^^^^^^^^
+        try:
+            return (
+                uri.split("://", 1)[1]
+                .split("/", 1)[0]
+                .split("@", 1)[1]
+                .split(".dfs.core.windows.net", 1)[0]
+            )
+
+        except IndexError:
+            return None
+
+    @staticmethod
+    def _get_azure_storage_account_key_az_cli(account_name: str) -> str:
+        az_cmd = [
+            "az",
+            "storage",
+            "account",
+            "keys",
+            "list",
+            "--output",
+            "json",
+            "--account-name",
+            account_name,
+        ]
+
+        cmd = az_cmd if sys.platform != "win32" else ["cmd", "/C", *az_cmd]
+
+        # [
+        #     {
+        #         "creationTime": "1970-01-01T00:00:00.000000+00:00",
+        #         "keyName": "key1",
+        #         "permissions": "FULL",
+        #         "value": "..."
+        #     },
+        #     {
+        #         "creationTime": "1970-01-01T00:00:00.000000+00:00",
+        #         "keyName": "key2",
+        #         "permissions": "FULL",
+        #         "value": "..."
+        #     }
+        # ]
+
+        return json.loads(subprocess.check_output(cmd))[0]["value"]
```

Comment:
This is ported from some recently added logic to the rust side from (https://github.com/pola-rs/polars/pull/20357), we need this here as the Python credential provider will override the one set by the Rust side if it is passed.